gitlabform/processors/project/job_token_scope_processor.py

```diff
@@ -0,0 +1,244 @@
+            if allowlist_updated:
+                # Refresh has no return but produces same result as project.job_token_scope.get()
+                # -> refreshes job_token_scope state with latest changes
+                job_token_scope.refresh()
+
+            info(
+                "Enforce enabled, removing projects no longer defined in config from allowlist"
+            )
+            self._remove_projects_from_allowlist(
+                project, job_token_scope, existing_allowlist, project_ids_to_allow
+            )
+
+    def _process_groups(
+        self,
+        job_token_scope: ProjectJobTokenScope,
+        groups_allowlist: List,
+        enforce: bool,
+    ):
+        if not groups_allowlist and enforce:
+            warning(
+                "Process will remove existing groups from allowlist, as none set in configuration"
+            )
+
+        existing_allowlist = job_token_scope.groups_allowlist.list(get_all=True)
+
+        group_ids_to_allow = self._get_target_group_ids_from_config(groups_allowlist)
+
+        allowlist_updated = False
+        if len(group_ids_to_allow) > 0:
+            allowlist_updated = self._add_groups_to_allowlist(
+                job_token_scope, existing_allowlist, group_ids_to_allow
+            )
+
+        if enforce:
+            if allowlist_updated:
+                # Refresh has no return but produces same result as project.job_token_scope.get()
+                # -> refreshes job_token_scope state with latest changes
+                job_token_scope.refresh()
+
+            info(
+                "Enforce enabled, removing groups no longer defined in config from allowlist"
+            )
+            self._remove_groups_from_allowlist(
+                job_token_scope, existing_allowlist, group_ids_to_allow
+            )
+
+    def _remove_groups_from_allowlist(
+        self,
+        job_token_scope: ProjectJobTokenScope,
+        existing_allowlist,
+        target_group_ids: List,
+    ):
+        allowlist_updated = False
+        group_ids_to_remove = self._get_ids_to_remove_from_allowlist(
+            existing_allowlist, target_group_ids
+        )
+        for group_id in group_ids_to_remove:
+            allowlist_updated = True
+            job_token_scope.groups_allowlist.delete(group_id)
+            verbose("Deleted group %s from allowlist", group_id)
+
+        if allowlist_updated:
+            verbose("Saving removed Groups allowlist changes")
+            job_token_scope.save()
+
+    @staticmethod
+    def _add_groups_to_allowlist(
+        job_token_scope, existing_allowlist, group_ids_listed_in_config
+    ):
+        allowlist_updated = False
+
+        for group_id in group_ids_listed_in_config:
+            if any(allowed.id == group_id for allowed in existing_allowlist):
+                # If already in allowlist, do nothing
+                verbose(f"{group_id} already in Groups allowlist")
+                pass
+
+            allowlist_updated = True
```

Comment:
@amimas it will only set to True if there is a group or project to add to the allowlist

If the only groups listed in the config are already on the allowlist, it won't save the scope

I wanted to keep it outside the if statement so we only save once and if necessary.

Reply:
Very glad you asked this @amimas, I added a test for it and discovered there was a bug in my logic!